SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/RPCService.cpp

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/RPCs/RPCService.h"
+
+namespace SpatialGDK
+{
+RPCService::RPCService(const FSubView& InRemoteSubView, const FSubView& InLocalAuthSubView)
+	: RemoteSubView(&InRemoteSubView)
+	, LocalAuthSubView(&InLocalAuthSubView)
+{
+}
+
+void RPCService::AddRPCQueue(FName QueueName, RPCQueueDescription&& Desc)
+{
+	Queues.Add(QueueName, MoveTemp(Desc));
+}
+
+void RPCService::AddRPCReceiver(FName ReceiverName, RPCReceiverDescription&& Desc)
+{
+	Receivers.Add(ReceiverName, MoveTemp(Desc));
+}
+
+void RPCService::AdvanceSenderQueues()
+{
+	for (const EntityDelta& Delta : LocalAuthSubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Update = Change.Update;
+				Ctx.Fields = Schema_GetComponentUpdateFields(Change.Update);
+
+				for (auto& QueueEntry : Queues)
+				{
+					if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
+					{
+						QueueEntry.Value.Sender->OnUpdate(Ctx);
+					}
+				}
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Fields = Schema_GetComponentDataFields(Change.CompleteUpdate.Data);
+
+				for (auto& QueueEntry : Queues)
+				{
+					if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
+					{
+						QueueEntry.Value.Sender->OnUpdate(Ctx);
+					}
+				}
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
```

Comment:
Why do we not check `|| Authority == 0` here?

Reply:
Queues need to write to an entity, so unlike receivers, they always need some authority, while receivers can receive in cases where they do not have any auth over an actor.